pandas/core/groupby/generic.py

```diff
@@ -535,24 +534,25 @@ def _transform_general(
             if isinstance(res, (ABCDataFrame, ABCSeries)):
                 res = res._values
 
-            indexer = self._get_index(name)
-            ser = klass(res, indexer)
-            results.append(ser)
+            results.append(klass(res, index=group.index))
 
         # check for empty "results" to avoid concat ValueError
         if results:
             from pandas.core.reshape.concat import concat
 
-            result = concat(results).sort_index()
+            concatenated = concat(results)
+            result = self._set_result_index_ordered(concatenated)
         else:
             result = self.obj._constructor(dtype=np.float64)
-
         # we will only try to coerce the result type if
         # we have a numeric dtype, as these are *always* user-defined funcs
         # the cython take a different path (and casting)
-        dtype = self._selected_obj.dtype
-        if is_numeric_dtype(dtype):
-            result = maybe_downcast_to_dtype(result, dtype)
+        if is_numeric_dtype(result.dtype):
+            common_dtype = np.find_common_type(
+                [self._selected_obj.dtype, result.dtype], []
+            )
+            if common_dtype is result.dtype:
```

Comment:
hmm, i think you can always do this no? (as the comparison is also done in the downcaster)

Reply:
I looked into this - this OP is the only place where this check is needed

in the OP we're doing 
``` python
_maybe_downcast_numeric(result, np.dtype('float64'))
```
with `result` being of `np.dtype('int64')`, so we unnecessarily upcast

i'm actually not sure why `self._selected_obj` is `np.dtype('float64')` b/c original dataframe is ints with some `NaN`

Reply:
my point is I don't think you need to check the common dtype at all here

Reply:
tried this (deleting the casting altogether, if i understood right)

two tests would be affected. In 
https://github.com/pandas-dev/pandas/blob/319a6d3f0d93c372bd53498168b957aaf5daa174/pandas/tests/groupby/transform/test_transform.py#L212-L218
`result` is now `float64` instead of `int64`, and same in
https://github.com/pandas-dev/pandas/blob/319a6d3f0d93c372bd53498168b957aaf5daa174/pandas/tests/groupby/transform/test_transform.py#L457-L462